File: python_solutions/616.add-bold-tag-in-string.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/add-bold-tag-in-string/description/
# Medium
class Solution:
    def addBoldTag(self, s: str, words: List[str]) -> str:
        bold = [False] * len(s)
        for w in words:
            start = s.find(w)
            while start != -1:
                for i in range(start, start + len(w)):
                    bold[i] = True
                start = s.find(w, start + 1)

        res = ""
        i = 0
        while i < len(s):
            if bold[i] == False:
                res += s[i]
                i += 1
            else:
                res += "<b>"
                while i < len(s) and bold[i] == True:
                    res += s[i]
                    i+=1
                res += "</b>"
        return res

File: python_solutions/616.add-bold-tag-in-string.cpp
// Containers
#include <deque>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "conf.d/say.h"

using namespace std;
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
class Solution {
public:
  string addBoldTag(string s, vector<string> &dict) {
    vector<int> bold(s.length(), false);
    for (auto &w : dict) {
      auto start = s.find(w);
      while (start != string::npos) {
        for (int i = start; i < start + w.length(); i++) bold[i] = true;
        start = s.find(w, start + 1);
      }
    }
    string res = "";
    int i = 0, n = s.length();
    while (i < n) {
      if (bold[i] == false)
        res.push_back(s[i++]);
      else {
        res += "<b>";
        while (i < n && bold[i] == true) res.push_back(s[i++]);
        res += "</b>";
      }
    }
    return res;
  }
};

int main() {
  Solution s;
  string t = "aaabbcc";
  vector<string> dict = {"aaa", "aab", "bc"};
  say(s.addBoldTag(t, dict));
}